azerothcore/azerothcore-wotlk
# fix(Core/Event): Rework STV Fishing Extravaganza

## Changes Proposed:
Rework STV Fishing Extravaganza.

## Issues Addressed:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9532

## SOURCE:
- Heavily inspired by VMangos 
- https://github.com/vmangos/core/blob/9d0aeb7d63b2989686f028b0c5c039b671dc9b25/src/scripts/world/npcs_special.cpp#L1434
- https://wowpedia.fandom.com/wiki/Stranglethorn_Fishing_Extravaganza?oldid=2373763
- https://github.com/vmangos/core/commit/4cd5ca80997a18de296129531158ae3cfb76898f
- https://github.com/vmangos/core/commit/4f739079a44423e31f7c2b895fc6f7068cb86e44

## Tests Performed:
- tested in game

## How to Test the Changes:
- Go to booty bay 
- `.event start 62` (spawns npc's - default time: 13:00)
- `.event start 90` (starts turn in quests, Riggle Bassbait should announce that the event has started - 14:00)
- `.event start 15` (pools are spawned - default time: 14:00)
- To see and accept quests, fishing skill >=150 is required
- Check Riggle Bassbait gossip and Jang before master angler
- Turn in master angler quest -> afterwards Jang quest becomes active, Riggle Bassbait has different gossip
- `.event stop 15` Riggle Bassbait should announce that pools are gone (default time: 16:00)
- `.event stop 90` Quests should become unavailable (default time: 17:00)
- `.event stop 62` All npc's despwan (default time: 17:00)
-  Restart server
- `.event start 62`
- `.event start 90`
- Riggle Bassbait master angler quest should remain disabled
- Riggle Bassbait gossip should remain "we already have winner"
- Event start announcement does not happen again
- After one day, all should be reset.

For manually resetting the event states:
1. Delete variables in table `worldstates` in character database for immediate retesting
```sql
DELETE FROM `worldstates` WHERE  `entry` IN (197,198,199,200);
```

## Known Issues and TODO List:
- World state ids ok?
- Noticed two issues (unrelated to this PR):

1. npc quest flag is not immediately updated, when condition changes. 
   This affects Fishbot 5000 and Jang, when event 90 starts. 
   Nearby players sometimes do not immediately see the quest giver indicator, but the quest are shown in the gossip menu and 
   can be accepted. After relog, quest giver indicator is correctly displayed.  (I only noticed this after converting the scripts to db
   with conditions, cpp script did always immediately show the quest giver indicator).

2. Event time is off by one hour.
    Example: start time should be 13:00 (set in db) but when I use `.event info 62` it shows that next active time is 14:00.
    Could be something similar as it was the case for hourly bells: 
    _[update] added tzset(); for localtime_r() in order to fix timezone/daylight saving time_
    https://github.com/azerothcore/azerothcore-wotlk/pull/9151

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
